RACF-KDFAES format fixes

solardiz opened the pull request:
This PR is to make the format a little bit more correct and reliable, maybe throw in some easy optimizations too (not done yet).

solardiz committed:
RACF-KDFAES format: Report tunable costs

solardiz committed:
RACF-KDFAES format: Benchmark only Raw, print correct algorithm detail

Previously, we'd wrongly see speedup for "Many salts" vs. "Only one salt",
which was actually an effect of the different costs in our test vectors.

solardiz committed:
RACF-KDFAES format: Stricter valid()

solardiz committed:
RACF-KDFAES format: Minor simplification

solardiz commented:
See also: #3486 #2511 #2510
cc: @bigendiansmalls

solardiz committed:
RACF-KDFAES format: Minor optimization

solardiz committed:
RACF-KDFAES format: Free memory in done()

bigendiansmalls commented:
Happy to test the changes and make sure nothing broke :) Will let you know. Holiday weekend so might take me a day.

solardiz committed:
RACF-KDFAES format: Even stricter valid(), and report costs differently

These changes are per valid ranges and naming seen on some slide decks.

solardiz committed:
RACF-KDFAES format: Swap the two test vectors

So that the first test vector, which we use for benchmark, closer resembles
what a slide deck gives as default (they give 8,50 and ours is 9,50 now).

solardiz committed:
RACF-KDFAES format: Specialize the HMAC implementation

Drop the "if" branches that are impossible in our usage here.

solardiz committed:
RACF-KDFAES format: Split the HMAC function in two

solardiz committed:
RACF-KDFAES format: Move HMAC key setup out of loops where possible

This provides a ~75% speedup.

solardiz committed:
RACF-KDFAES format: Embed hash output in input message union

solardiz committed:
RACF-KDFAES format: Avoid copying hash in the random access loop

solardiz commented:
There's finally one commit here (building upon preceding commits) that makes this ~30% faster, but it's still very slow indeed, even when using SHA-NI via OpenSSL (I've confirmed that it does for me).

A further step could be to implement SIMD for the nested loop, which shouldn't be too hard and would cover most of the running time. But I probably won't bother now.

solardiz commented:
> There's finally one commit here (building upon preceding commits) that makes this ~30% faster,

I had missed part of the optimization there. Corrected. Now this one commit makes the format ~75% faster. Along with the rest of the changes here, we're ~80% faster.

solardiz committed:
RACF-KDFAES format: Unroll HMAC pad XOR loops

solardiz committed:
RACF-KDFAES format: Copy the HMAC implementation from hmac_sha_plug.c

solardiz committed:
RACF-KDFAES format: Use mostly do-while loops

solardiz commented:
Having worked on these changes, I now have a better idea of what this KDFAES thing is. It's modeled after scrypt, where there are 3 steps: key expansion, memory-hard loop, and calculation of digest of the final state.

1. In scrypt, key expansion is done with PBKDF2 at 1 iteration but a derived key length of like 1 KiB. It produces just the initial block B. In KDFAES, it's done with an algorithm extremely similar to but not quite PBKDF2, the iteration count is high, and it produces the full memory content, which in scrypt would be V.

2. The memory-hard loop is similar to scrypt SMix's second loop, except that besides random reads (similar to scrypt's) it writes over memory (scrypt's only reads), which it does sequentially. The first loop (which would fill V given B) is missing - or alternatively we can say that it's replaced by the PBKDF2-alike's outer loop (over derived key blocks).

3. The final state digest calculation in scrypt is done with PBKDF2 at 1 iteration over the final block. Here it's done over the full memory content, and again with high iteration count. The algorithm _may_ be literally PBKDF2 (I didn't verify, but I'm also not aware of this one deviating from proper PBKDF2).

The initial key expansion in scrypt is quick, but in KDFAES it takes almost all its running time because the complexity is quadratic. The would-be memory-hard portion only goes over memory once (fine on its own) and unlike the key expansion it doesn't use iterations, making it _relatively_ quick. So this isn't truly a memory-hard algorithm. The final digest calculation doesn't suffer from quadratic complexity because the large input (full memory content) is hashed first, so even despite of the high iteration count it's _relatively_ quick.

So we may try calling into our shared PBKDF2 code for step 3, but this isn't going to affect overall performance much. What will make more of a difference is calling into our SIMD SHA-256 for step 1.

solardiz committed:
RACF-KDFAES format: Specialize the HMAC for key length <= 32

solardiz committed:
RACF-KDFAES format: Avoid dynamic memory allocation (use stack)

solardiz commented:
> Happy to test the changes and make sure nothing broke :) Will let you know. Holiday weekend so might take me a day.

I've just merged, but of course please do test and let us know when you can. Also I suspect that adding support for length > 8 may require a revert of some of the HMAC code specialization here. That's fine.

solardiz commented:
Just because I was documenting my findings in this PR, I'll also add in here today's Kaspersky SecureList blog post on the topic:

https://securelist.com/zos-mainframe-pentesting-resource-access-control-facility/116873/

There's a lot, but one piece I'd like to quote is:

> Encrypting a password phrase hash with KDFAES shares many similarities with encrypting a password hash. According to public sources, the primary difference lies in the key used during the second stage. For passwords, data derived from DES-encrypting the username was used, while for a password phrase, its SHA256 hash is used. During our analysis, we could not determine the exact password phrase hashing process – specifically, whether padding is involved, if a secret key is used, and so on.
> 
> Additionally, when using a password phrase, the PHRASE and PHRASEX fields instead of PASSWORD and PWDX, respectively, store the final hash, with the PHRASEX value having a similar structure.

So to support passwords (phrases) longer than 8, apparently the following would be needed:
1. An extension to `racf2john` to extract more fields.
2. Figure out the algorithm and implement it.

solardiz commented:
There's also a hashcat PR implementing this algorithm (for length up to 8 like we have) https://github.com/hashcat/hashcat/pull/3940 by @c0rs
